i686-elf-grub 2.06, x86_64-elf-grub 2.06 (new formulae)

rabarbra opened the pull request:
<!-- Use [x] to mark item done, or just click the checkboxes with device pointer -->

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

rabarbra commented:
The CI pipeline is failing because of CI/Linux job.
The CI/Linux is failing because of this:

```==> brew audit --formula i686-elf-grub --online --new
==> FAILED
Full audit --formula i686-elf-grub --online --new output
  i686-elf-grub
    * The homepage URL https://www.gnu.org/software/grub/ is not reachable
    * Binaries built for a non-native architecture were installed into i686-elf-grub's prefix.
      The offending files are:
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-boot.image	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-boot_hybrid.image	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-cdboot.image	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-diskboot.image	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-kernel.exec	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-lnxboot.image	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-lzma_decompress.image	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/i686-elf-pxeboot.image	(i386)
        /home/linuxbrew/.linuxbrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/kernel.img	(i386)
  Error: 2 problems in 1 formula detected.
  ```


The CI pipeline is showing failures in the Linux job due to two issues:

1. Homepage unreachable: This appears to be a false positive as https://www.gnu.org/software/grub/ is accessible and passes the check in the macOS jobs.

2. Non-native architecture binaries: The audit is flagging i386 binaries in the package. However, this is expected and necessary behavior for this formula as it's a cross-compiler toolchain specifically designed to produce i386/i686 binaries. And brew audit is again passing macOS jobs, while prefix still contains non-native binaries.

Would appreciate guidance on how to best handle these CI checks.

botantony commented:
This is a common problem with gnu.org/nongnu.org homepages (#206757). Try to use savannah.gnu.org if it has a homepage for these programs

botantony commented:
also can you please add at least one mirror for each formula?

rabarbra commented:
Thank you for quick reaction!

GRUB tools are used to create and work with bootable images for specific architectures. 

Tools like grub-mkrescue, grub-mkimage, and grub-mkstandalone embed architecture-specific software during their compilation process. This is controlled through target-specific variables:
```TARGET_CC=#{target}-gcc
TARGET_OBJCOPY=#{target}-objcopy  
TARGET_STRIP=#{target}-strip
TARGET_NM=#{target}-nm
TARGET_RANLIB=#{target}-ranlib
```

I have these in my formulas.

There is no universal version of these tools that can produce bootable images for any architecture. Instead:

To create bootable images for i686 (32-bit) systems, you need i686-elf-grub-mkrescue compiled with tools for i686 architecture
To create bootable images for x86_64 (64-bit) systems, you need x86_64-elf-grub-mkrescue

When installing GRUB, you typically only install the tools for the specific architecture you're working with. This approach leverages the cross-compiler toolchains already available in Homebrew (i686-elf-binutils and x86_64-elf-binutils for example), allowing you to build boot images targeted at specific architectures regardless of your host system.

I will change homepage to https://savannah.gnu.org/projects/grub later today and add this mirror: https://mirrors.ocf.berkeley.edu/gnu/grub/grub-2.06.tar.xz.
But formulas will still not pass the Linux CI job because to brew audit detecting non-native binaries. Question is if we can omit this check. And why does it pass all the MacOs CIs, while there are still non-native binaries in formula's prefix. Why this test is different for MacOs and Linux?

botantony commented:
macOS CI is kinda strange (maybe because macOS uses Clang vs. Linux uses GCC) but I can say for sure that Homebrew would never ever accept non-native binaries

rabarbra commented:
But these binaries are not executables, they are images for specific architectures, used by GRUB tools to work with images and GRUB itself compiled for these architectures, needed to create images.

Problem is how brew audit defines if it is executable binary in Linux. It supposed to skip not executables (see this line: https://github.com/Homebrew/brew/blob/b4dbbf19a21352dd5ed0748e32e6c3b7f15e800c/Library/Homebrew/extend/os/linux/keg_relocate.rb#L76).

It happens like this in MacOs: ruby-macho gem just returns ":dunno" for these files (https://github.com/Homebrew/brew/blob/master/Library/Homebrew/os/mac/mach.rb).

But for Linux the test is following: https://github.com/Homebrew/brew/blob/b4dbbf19a21352dd5ed0748e32e6c3b7f15e800c/Library/Homebrew/os/linux/elf.rb#L91.
It only tests if it is elf file with ET_EXEC flag in ELF header's e_type field. But file like this is not necessary executable. GRUB's make install gives only read permissions for these files, not execute.

gromgit commented:
Try adding `lib/i686-elf/**` to `audit_exceptions/mismatched_binary_allowlist.json`.

fxcoudert commented on Formula/x/x86_64-elf-grub.rb:
```diff
@@ -0,0 +1,65 @@
+  sha256 "b79ea44af91b93d17cd3fe80bdae6ed43770678a9a5ae192ccea803ebb657ee1"
+  license "GPL-3.0-or-later"
+
+  depends_on "autoconf" => :build
```
Since the configure script is shipped, it should built without autoconf/automake

rabarbra replied:
You are right, I'll remove these.

fxcoudert commented on Formula/x/x86_64-elf-grub.rb:
```diff
@@ -0,0 +1,65 @@
+
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "gcc@11" => [:build, :test]
```
Does it really need GCC? And only one specific (outdated) version?

rabarbra replied:
I added version 11 here for a reason, but I forgot why. I'll try to find this case again and take closer look, if this can be changed.

fxcoudert commented on Formula/x/x86_64-elf-grub.rb:
```diff
@@ -0,0 +1,65 @@
+  depends_on "automake" => :build
+  depends_on "gcc@11" => [:build, :test]
+  depends_on "help2man" => :build
+  depends_on "make" => :build
```
Does it only build with GNU make? It seems unlikely.

fxcoudert commented on Formula/x/x86_64-elf-grub.rb:
```diff
@@ -0,0 +1,65 @@
+  depends_on "texinfo" => :build
+  depends_on "x86_64-elf-binutils" => :build
+  depends_on "x86_64-elf-gcc" => [:build, :test]
+  depends_on "gettext"
```
Is this really a runtime dependence? Not build-only?

rabarbra replied:
Yes, it is:
```
➜  ~ otool -L /opt/homebrew/bin/i686-elf-grub-file
/opt/homebrew/bin/i686-elf-grub-file:
        /opt/homebrew/opt/gettext/lib/libintl.8.dylib (compatibility version 13.0.0, current version 13.2.0)
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 3208.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1351.0.0)
➜  ~ otool -L /opt/homebrew/bin/i686-elf-grub-glue-efi
/opt/homebrew/bin/i686-elf-grub-glue-efi:
        /opt/homebrew/opt/gettext/lib/libintl.8.dylib (compatibility version 13.0.0, current version 13.2.0)
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 3208.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1351.0.0)
➜  ~ otool -L /opt/homebrew/bin/i686-elf-grub-mkimage
/opt/homebrew/bin/i686-elf-grub-mkimage:
        /opt/homebrew/opt/xz/lib/liblzma.5.dylib (compatibility version 12.0.0, current version 12.4.0)
        /opt/homebrew/opt/gettext/lib/libintl.8.dylib (compatibility version 13.0.0, current version 13.2.0)
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 3208.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1351.0.0)
 ```

fxcoudert commented on Formula/x/x86_64-elf-grub.rb:
```diff
@@ -0,0 +1,65 @@
+        --libdir=#{lib}/#{target}
+        --with-platform=pc
+        --program-prefix=#{target}-
+        TARGET_CC=#{target}-gcc
```
I'm surprised because GNU tools are usually able to detect these target tools if the correct `-build`, `-host` and `-target` arguments (one or more of them, probably `-target` in this case) is provided.

botantony commented:
I still don't understand why we have to split this formula. Why not build 2 different targets in one formula?

rabarbra commented:
This problem with homepage again in CI. Even with https://savannah.gnu.org/projects/grub as homepage.
But only in single job -  macOS 14-x86_64. All other jobs (Linux, macOS arm64 jobs and  macOS 13-x86_64) are fine.

```
==> brew audit --formula i686-elf-grub --online --new
==> FAILED
Full audit --formula i686-elf-grub --online --new output
  i686-elf-grub
    * The homepage URL https://savannah.gnu.org/projects/grub is not reachable
  Error: 1 problem in 1 formula detected.
```

rabarbra commented:
The following explains mtools and freetype dependencies for x86_64 formula:

```
➜  ~ x86_64-elf-grub-mkrescue iso -o test.iso
GNU xorriso 1.5.6 : RockRidge filesystem manipulator, libburnia project.

Drive current: -outdev 'stdio:test.iso'
Media current: stdio file, overwriteable
Media status : is blank
Media summary: 0 sessions, 0 data blocks, 0 data,  128g free
Added to ISO image: directory '/'='/var/folders/wq/6r2mrd8x103g0lmswg7m3cc00000gn/T/grub.6mLtw5'
xorriso : UPDATE :     331 files added in 1 seconds
Added to ISO image: directory '/'='/Users/.../iso'
xorriso : UPDATE :     335 files added in 1 seconds
xorriso : WARNING : EFI boot equipment is provided but no directory /EFI/BOOT
xorriso : WARNING : will emerge in the ISO filesystem. A popular method to
xorriso : WARNING : prepare a USB stick on MS-Windows relies on having in the
xorriso : WARNING : ISO filesystem a copy of the EFI System Partition tree.
ISO image produced: 7278 sectors
Written to medium : 7278 sectors at LBA 0
Writing to 'stdio:test.iso' completed successfully.

➜  ~ brew uninstall mtools --ignore-dependencies
Uninstalling /opt/homebrew/Cellar/mtools/4.0.48... (69 files, 583.2KB)
➜  ~ x86_64-elf-grub-mkrescue iso -o test.iso
x86_64-elf-grub-mkrescue: error: `mformat` invocation failed
.
➜  ~ otool -L /opt/homebrew/Cellar/x86_64-elf-grub/2.06/bin/x86_64-elf-grub-mkfont
/opt/homebrew/Cellar/x86_64-elf-grub/2.06/bin/x86_64-elf-grub-mkfont:
        /opt/homebrew/opt/freetype/lib/libfreetype.6.dylib (compatibility version 27.0.0, current version 27.2.0)
        /opt/homebrew/opt/gettext/lib/libintl.8.dylib (compatibility version 13.0.0, current version 13.3.0)
        /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 3208.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1351.0.0)
➜  ~ i686-elf-grub-mkrescue iso -o test2.iso
GNU xorriso 1.5.6 : RockRidge filesystem manipulator, libburnia project.

Drive current: -outdev 'stdio:test2.iso'
Media current: stdio file, overwriteable
Media status : is blank
Media summary: 0 sessions, 0 data blocks, 0 data,  128g free
Added to ISO image: directory '/'='/var/folders/wq/6r2mrd8x103g0lmswg7m3cc00000gn/T/grub.098Ctx'
xorriso : UPDATE :     329 files added in 1 seconds
Added to ISO image: directory '/'='/Users/psimonenko/iso'
xorriso : UPDATE :     333 files added in 1 seconds
xorriso : NOTE : Copying to System Area: 512 bytes from file '/opt/homebrew/Cellar/i686-elf-grub/2.06/lib/i686-elf/grub/i386-pc/boot_hybrid.img'
ISO image produced: 4163 sectors
Written to medium : 4163 sectors at LBA 0
Writing to 'stdio:test2.iso' completed successfully.

➜  ~ brew info mtools
==> mtools: stable 4.0.48 (bottled)
Tools for manipulating MSDOS files
https://www.gnu.org/software/mtools/
Conflicts with:
  multimarkdown (because both install `mmd` binaries)
Not installed
Bottle Size: 179.2KB
Installed Size: 581.8KB
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/m/mtools.rb
License: GPL-3.0-or-later
==> Analytics
install: 261 (30 days), 760 (90 days), 2,173 (365 days)
install-on-request: 251 (30 days), 750 (90 days), 2,163 (365 days)
build-error: 0 (30 days)
```

fxcoudert commented:
Hi @rabarbra these new formulas you introduced are now failing in the latest update. Can you please have a look? https://github.com/Homebrew/homebrew-core/pull/209489

rabarbra commented:
Hi! 
`../../grub-core/extra_deps.lst` is needed. The file can be empty - should just present. I checked locally - in this case everything is compiling and working.

```
➜ ls
00_header         garbage-gen        grub-glue-efi.1     grub-mkfont             grub-mkrescue        grub-render-label       grub-syslinux2cfg.1             libgrub_a_init.c                 tests
10_xnu            grub-bios-setup    grub-install        grub-mkfont.1           grub-mkrescue.1      grub-render-label.1     grub_fstest-grub_fstest_init.o  libgrub_a_init.lst               util
25_bli            grub-bios-setup.8  grub-install.8      grub-mkimage            grub-mkstandalone    grub-script-check       grub_fstest.pp                  libgrubgcry.a
30_os-prober      grub-core          grub-kbdcomp        grub-mkimage.1          grub-mkstandalone.1  grub-script-check.1     grub_fstest_init.c              libgrubkern.a
30_uefi-firmware  grub-editenv       grub-kbdcomp.1      grub-mklayout           grub-mount           grub-set-default        grub_fstest_init.lst            libgrubmods.a
40_custom         grub-editenv.1     grub-macbless       grub-mklayout.1         grub-mount.1         grub-set-default.8      grub_mount-grub_fstest_init.o   libgrubmods_a-grub_script.tab.o
41_custom         grub-file          grub-macbless.8     grub-mknetdir           grub-ofpathname      grub-shell              grub_script.tab.c               libgrubmods_a-grub_script.yy.o
config-util.h     grub-file.1        grub-menulst2cfg    grub-mknetdir.1         grub-ofpathname.8    grub-shell-luks-tester  grub_script.tab.h               libgrubmods_a-libgrub_a_init.o
config.h          grub-fs-tester     grub-menulst2cfg.1  grub-mkpasswd-pbkdf2    grub-probe           grub-shell-tester       grub_script.yy.c                Makefile
config.log        grub-fstest        grub-mkconfig       grub-mkpasswd-pbkdf2.1  grub-probe.8         grub-sparc64-setup      grub_script.yy.h                po
config.status     grub-fstest.1      grub-mkconfig.8     grub-mkrelpath          grub-reboot          grub-sparc64-setup.8    include                         stamp-h
docs              grub-glue-efi      grub-mkconfig_lib   grub-mkrelpath.1        grub-reboot.8        grub-syslinux2cfg       libgrub.pp                      stamp-h1
➜ pwd
/Users/.../Prog/pr/grub-2.12/build
➜ cat ../../grub-core/extra_deps.lst
➜ ls ../../grub-core/extra_deps.lst
../../grub-core/extra_deps.lst
```

rabarbra commented:
But I would also double-check build time and run-time dependencies for 2.12 first.
At least gawk is needed for build time. So, if you fix extra_deps.lst, you will meet a problem with missing gawk during compilation, if you don't add it to the formula.

rabarbra commented:
And gawk is being used as awk, so probably this is needed for gawk from homebrew:

```
GNU "awk" has been installed as "gawk". If you need to use it as "awk", you can add a "gnubin" directory to your PATH from your ~/.bashrc and/or ~/.zshrc like:     PATH="$HOMEBREW_PREFIX/opt/gawk/libexec/gnubin:$PATH"
```

(https://formulae.brew.sh/formula/gawk#default)

rabarbra commented:
I have to work now, but maybe tomorrow or on Sunday I can try to create a commit / pull request if nobody does it before.

rabarbra commented:
This commit fixes the issue with extra-deps.lst, but it was added after the release of version 2.12: https://git.savannah.gnu.org/cgit/grub.git/commit/?id=b835601c7639ed1890f2d3db91900a8506011a8e

rabarbra commented:
https://github.com/Homebrew/homebrew-core/pull/210240
https://github.com/Homebrew/homebrew-core/pull/210242
